Sources/Basics/TemporaryFile.swift

```diff
@@ -0,0 +1,111 @@
+///           return value for the `withTemporaryDirectory` function.
+///           The cleanup block should be called when the temporary directory is no longer needed.
+///
+/// - Throws: `MakeDirectoryError` and rethrows all errors from `body`.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public func withTemporaryDirectory<Result>(
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    _ body: (AbsolutePath, @escaping (AbsolutePath) -> Void) async throws -> Result
+) async throws -> Result {
+    let temporaryDirectory = try createTemporaryDirectory(dir: dir, prefix: prefix)
+    return try await body(temporaryDirectory) { path in
+        _ = try? FileManager.default.removeItem(atPath: path.pathString)
+    }
+}
+
+/// Creates a temporary directory and evaluates a closure with the directory path as an argument.
+/// The temporary directory will live on disk while the closure is evaluated and will be deleted afterwards.
+///
+/// This function is basically a wrapper over posix's mkdtemp() function.
+///
+/// - Parameters:
+///     - dir: If specified the temporary directory will be created in this directory otherwise environment
+///            variables TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env
+///            variables are set, dir will be set to `/tmp/`.
+///     - prefix: The prefix to the temporary file name.
+///     - removeTreeOnDeinit: If enabled try to delete the whole directory tree otherwise remove only if its empty.
+///     - body: A closure to execute that receives the absolute path of the directory as an argument.
+///             If `body` has a return value, that value is also used as the
+///             return value for the `withTemporaryDirectory` function.
+///
+/// - Throws: `MakeDirectoryError` and rethrows all errors from `body`.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public func withTemporaryDirectory<Result>(
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    removeTreeOnDeinit: Bool = false ,
+    _ body: (AbsolutePath) async throws -> Result
+) async throws -> Result {
+    try await withTemporaryDirectory(dir: dir, prefix: prefix) { path, cleanup in
+        defer { if removeTreeOnDeinit { cleanup(path) } }
+        return try await body(path)
+    }
+}
+
+private func createTemporaryDirectory(dir: AbsolutePath?, prefix: String) throws -> AbsolutePath {
+    // Construct path to the temporary directory.
+    let templatePath = try AbsolutePath(prefix + ".XXXXXX", relativeTo: determineTempDirectory(dir))
+    
+    // Convert templatePath to a C style string terminating with null char to be an valid input
+    // to mkdtemp method. The XXXXXX in this string will be replaced by a random string
+    // which will be the actual path to the temporary directory.
+    var template = templatePath.pathString.utf8CString.map { $0 }
+    
+    if TSCLibc.mkdtemp(&template) == nil {
```

Comment:
this is where things get real :) iiuc `TSCLibc.mkdtemp` is hard wired to use the local file system. we should do a couple of things here:
1. change `createTemporaryDirectory` to also take a `FileSystem` to operate on (and wire it up in the call sites)
2. replace TSCLibc.mkdtemp with something more robust, potentially make it new API of `FileSystem`, so that we can have flexibility around testing and mocking it. The one in `LocalFileSystem` (from TSC) can continue to use `TSCLibc.mkdtemp` as its underlying implementation

Reply:
I found out that we do have an API inside `FileSystem` protocol which can be used for creating directories. So i changed it to use it and also saved previous behavior with generating random suffix with length of 6 numbers.
Now it is using `FileSystem` API which can be changed if it is needed for testing and mocking